### ISSUE TITLE
Proposal: expand Type functionality

### ISSUE THREAD

**Issue by phillipskevin** [canjs/can-define#287]
**tldr**

To make `can-define` more useful and add some long-requested functionality, we will:

* add support for things like `prop: String`
* expand the available `Type`s
* add strict `Type`s
* add support for `required` properties

### Allow shorthand with built-in Constructors

Instead of `type: "string"`, `type: "number"`, etc we should add support for using built-in constructors with `Type` like:

```
var VM = can.DefineMap.extend({
  prop: String
});
```

### More Types

Also, we should add additional types. This is not a definitive list, just some things that have been requested and others that exist in other languages:

* float
* integer
* char
* set
* enum / union
* tuple

### Strict Types

We could also create _strict_ types that would throw errors if set to a value of a different type. This might look something like:

```js
import { StrictString } from 'can-define-strict-types';
import DefineMap from 'can-define/map/map';

const VM = DefineMap.extend({
    prop: StrictString
});
```

This may also be called `MustString` or something like that... we can bikeshed on naming later.

### Required Values

Tangentially related to Types, people have also asked for support for [required](https://github.com/canjs/can-define/issues/62) [properties](https://github.com/canjs/can-define/issues/173). While expanding the Type behavior, we would also add support for this.

**Comment by phillipskevin** [comment]
@christopherjbaker pointed out that we should take a look at https://www.npmjs.com/package/prop-types for inspiration.
